### PR TITLE
Added the support to block the API calls based on Client IP and email domains. 

### PR DESCRIPTION
**Pull request by susmithp**
We need to have a YAML file name 'prohibited.yaml" with following keys.

BANNED_DOMAINS: (comma separated domain names, wildcard asterisk * is acceptable )
BANNED_IP: (Comma separated IPs, it can be single IP, Subnet, or range of IPs)

e.g. 
BANNED_DOMAINS: xyz.com, *abc*, *abcd.com
BANNED_IP: 10.10.1.1, 10.1.1.1/24, 10.10.1.1-10.10.1.100

Newly added: base/ipfilter.go
This package will read from Yaml file, and store it in a data structure .  For a HTTP request, it will try to search the Client IP in the data structure to find the match, and will flag the HTTP request accordingly

### CODE REVIEW

**Comment by arunkoshy** [comment]
this is looking good @susmithp -- we will let this cook for a bit and merge soon.

File: base/ipfilter.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+var ProhibitedIPs Prohibited
+var ProhibitedDomains string
+
+func InitProhibitedIPs() {
+	ProhibitedIPs.ips = make(map[string]bool)
+	config := viper.New()
+	config.SetConfigName("prohibited")
+	config.SetConfigType("yaml")
+	config.AddConfigPath("/usr/local/production/config/")
```
Comment (vejmarie) [inline base/ipfilter.go:37]: Can you make the config path an environment variable or external to the source code accessible path ?

File: base/ipfilter.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+	err := config.ReadInConfig()
+	if err != nil {
+		Zlog.Errorf("Falied to Initialise the Prohibited domain data: %s", err.Error())
```
Comment (vejmarie) [inline base/ipfilter.go:41]: Got a typo to Falied -> Failed. Why putting capital letter to Initialize and Prohibited ?

Comment (arunkoshy) [reply base/ipfilter.go:41]: @susmithp bits we can pick up through spell-check -- will suggest plugins offline --

-- corrected versions for you for ipfilter.go --

Zlog.Errorf("Failed to initialize prohibited domains list: %s", err.Error())
Zlog.Infof("Config file changed")
Zlog.Infof("User email domain [%s] is safe to procced", userDomain)

File: base/ipfilter.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		return true
+	}
+	blockedDomains := strings.Split(strings.ReplaceAll(ProhibitedDomains, " ", ""), ",")
+	at := strings.LastIndex(userEmail, "@")
+	userDomain := userEmail[at+1:]
+	Zlog.Infof("Verifying if email Domain[%s] belongs to the Banned domains.", userDomain)
+	for _, bdomain := range blockedDomains {
+		domain := strings.ReplaceAll(bdomain, ".", `\.`)
+		domain = strings.ReplaceAll(domain, "*", ".*")
+		Zlog.Debugf("Domains:[%s]:[%s]", bdomain, domain)
```
Comment (arunkoshy) [inline base/ipfilter.go:136]: @susmithp could you either drop Debugf or convert to Infof based on our offline discussion.

File: gateway/server.go [hunk 4]
```diff
@@ -303,6 +306,11 @@ func home(w http.ResponseWriter, r *http.Request) {
 	// And need to re route the end user to an end of session
 	switch head {
 	case "get_server_models":
+		base.Zlog.Infof("GET SERVER MODEL")
+		if base.ValidateClientIP(r) == false {
+			http.Error(w, "Service is not available", 401)
```
Comment (arunkoshy) [inline gateway/server.go:311]: @susmithp I think we need a more easily identifiable or distinguishable log message here i.e. via Zlog for easy retrieval of all forbidden domain / ip blocks and then also some of the other more broader error handling issues discussed by @vejmarie

**Review by arunkoshy — CHANGES_REQUESTED** [review]

File: base/ipfilter.go (hunk 1 above)
Comment (susmithp) [reply base/ipfilter.go]: @vejmarie I have corrected the typo, and changed the path to ENV variable.

File: base/ipfilter.go (hunk 2 above)
Comment (susmithp) [reply base/ipfilter.go]: @arunkoshy I have corrected the typo, please re-review.

File: base/ipfilter.go (hunk 3 above)
Comment (susmithp) [reply base/ipfilter.go]: @arunkoshy I have corrected it.

File: gateway/server.go (hunk 4 above)
Comment (susmithp) [reply gateway/server.go:310]: @arunkoshy I added that line just for debugging, and have removed it now.

File: base/ipfilter.go (hunk 3 above)
Comment (arunkoshy) [reply base/ipfilter.go]: LGTM now.

**Review by arunkoshy — APPROVED** [review]
This is looking good proviso re-test goes well ( observation time on staging ).

**Comment by arunkoshy** [comment]
@vejmarie this is ready to merge. Please re-review.

**Review by vejmarie — APPROVED** [review]
Missing examples, but let's approve it to move forward

**Comment by arunkoshy** [comment]
nicely done @susmithp ! Good job !

**Comment by arunkoshy** [comment]
@vejmarie yes, will provide example on my next PR which will open soon.

**Pull request merged**